### ISSUE TITLE
test exception when upgrading to 3.4

### ISSUE THREAD

**Issue by andyczerwonka** [etorreborre/specs2#359]
I upgraded my Play application to the latest specs2 3.4. Here's a look at my `build.sbt` configuration.
```
name := "freeze"

version := "1.0-SNAPSHOT"

lazy val root = (project in file(".")).enablePlugins(PlayScala)

scalaVersion := "2.11.6"

resolvers += "Scalaz Bintray Repo" at "http://dl.bintray.com/scalaz/releases"

libraryDependencies ++= Seq(
  cache,
  ws,
  "org.mindrot"  % "jbcrypt"   % "0.3m",
  "javax.mail" % "mail" % "1.4.5",
  "org.specs2" %% "specs2-core" % "3.4-scalaz-7.0.6" % "test",
  "org.specs2" %% "specs2-matcher-extra" % "3.4-scalaz-7.0.6" % "test",
  "org.specs2" %% "specs2-analysis" % "3.4-scalaz-7.0.6" % "test",
  "org.specs2" %% "specs2-mock" % "3.4-scalaz-7.0.6" % "test",
  "org.specs2" %% "specs2-junit" % "3.4-scalaz-7.0.6" % "test",
  "org.mongodb" %% "casbah" % "2.8.0"
)

scalacOptions in Test ++= Seq("-Yrangepos")
```
When I run the following test:
``` scala
@RunWith(classOf[JUnitRunner])
class PlanPersistenceSpec extends Specification with BeforeAfterEach {

  def before = db("plans").remove(MongoDBObject.empty)
  def after = db("plans").remove(MongoDBObject.empty)

  val payload = """{"name":"Fiver","age":4,"role":null}"""
  val userId = new ObjectId("551d5cf3985f20e469a1402a")

  "The Plan DAO" should {

    "create a plan" in {
      val plan = Plan.create("Some name", userId, Json.parse(payload)).get
      plan.id must not be null
      plan.name mustEqual "Some name"
      Json.stringify(plan.payload) mustEqual payload
    }

  }

}
```
```
[freeze] $ testOnly model.PlanPersistenceSpec
[info] Compiling 1 Scala source to /home/aczerwon/work/freeze/target/scala-2.11/test-classes...
[error] 
[error] java.lang.reflect.InvocationTargetException
[error]   caused by java.lang.AbstractMethodError: model.PlanPersistenceSpec.org$specs2$matcher$FuturezBaseMatchers$_setter_$org$specs2$matcher$FuturezBaseMatchers$$DefaultDaemonThreadFactory_$eq(Ljava/util/concurrent/ThreadFactory;)V
[error] 
[error] STACKTRACE
[error]   sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[error]   sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
[error]   sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[error]   java.lang.reflect.Constructor.newInstance(Constructor.java:526)
[error]   org.specs2.reflect.Classes$$anonfun$org$specs2$reflect$Classes$$createInstanceForConstructor$1$$anonfun$apply$11.apply(Classes.scala:105)
[error]   org.specs2.control.ActionT$$anonfun$safe$1.apply(ActionT.scala:89)
[error]   org.specs2.control.ActionT$$anonfun$reader$1$$anonfun$apply$6.apply(ActionT.scala:80)
[error]   org.specs2.control.Status$.safe(Status.scala:100)
[error]   org.specs2.control.StatusT$$anonfun$safe$1.apply(StatusT.scala:62)
[error]   org.specs2.control.StatusT$$anonfun$safe$1.apply(StatusT.scala:62)
[error]   scalaz.syntax.ToApplicativeOps$$anon$1.self$lzycompute(ApplicativeSyntax.scala:30)
[error]   scalaz.syntax.ToApplicativeOps$$anon$1.self(ApplicativeSyntax.scala:30)
[error]   scalaz.syntax.ToApplicativeOps$ApplicativeIdV$$anonfun$point$1.apply(ApplicativeSyntax.scala:34)
[error]   scalaz.WriterTApplicative$$anonfun$point$1.apply(WriterT.scala:266)
[error]   scalaz.WriterTApplicative$$anonfun$point$1.apply(WriterT.scala:266)
[error]   scalaz.effect.IO$$anonfun$apply$19$$anonfun$apply$20.apply(IO.scala:137)
[error]   scalaz.effect.IO$$anonfun$apply$19$$anonfun$apply$20.apply(IO.scala:137)
[error]   scalaz.FreeFunctions$$anonfun$return_$1.apply(Free.scala:288)
[error]   scalaz.FreeFunctions$$anonfun$return_$1.apply(Free.scala:288)
[error]   scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:33)
[error]   scalaz.Free$$anonfun$run$1.apply(Free.scala:174)
[error]   scalaz.Free$$anonfun$run$1.apply(Free.scala:174)
[error]   scalaz.Free.go2$1(Free.scala:117)
[error]   scalaz.Free.go(Free.scala:120)
[error]   scalaz.Free.run(Free.scala:174)
[error]   scalaz.effect.IO$class.unsafePerformIO(IO.scala:23)
[error]   scalaz.effect.IOFunctions$$anon$5.unsafePerformIO(IO.scala:221)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1$$anonfun$3.apply(Classes.scala:37)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1$$anonfun$3.apply(Classes.scala:36)
[error]   scala.collection.immutable.List.map(List.scala:273)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1.apply(Classes.scala:36)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1.apply(Classes.scala:29)
[error]   scala.Function1$$anonfun$andThen$1.apply(Function1.scala:55)
[error]   org.specs2.control.Status$class.fold(Status.scala:30)
[error]   org.specs2.control.Ok.fold(Status.scala:95)
[error]   org.specs2.control.Status$class.flatMap(Status.scala:48)
[error]   org.specs2.control.Ok.flatMap(Status.scala:95)
[error]   org.specs2.control.Status$class.map(Status.scala:45)
[error]   org.specs2.control.Ok.map(Status.scala:95)
[error]   org.specs2.control.StatusT$$anonfun$map$1.apply(StatusT.scala:16)
[error]   org.specs2.control.StatusT$$anonfun$map$1.apply(StatusT.scala:16)
[error]   scalaz.WriterT$$anonfun$map$1.apply(WriterT.scala:48)
[error]   scalaz.WriterT$$anonfun$map$1.apply(WriterT.scala:48)
[error]   scalaz.effect.IO$$anonfun$map$1$$anonfun$apply$8.apply(IO.scala:57)
[error]   scalaz.effect.IO$$anonfun$map$1$$anonfun$apply$8.apply(IO.scala:56)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:64)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:64)
[error]   scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$2.apply(Free.scala:72)
[error]   scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$2.apply(Free.scala:72)
[error]   scalaz.Free.resume(Free.scala:80)
[error]   scalaz.Free.go2$1(Free.scala:116)
[error]   scalaz.Free.go(Free.scala:120)
[error]   scalaz.Free.run(Free.scala:174)
[error]   scalaz.effect.IO$class.unsafePerformIO(IO.scala:23)
[error]   scalaz.effect.IOFunctions$$anon$5.unsafePerformIO(IO.scala:221)
[error]   org.specs2.runner.SbtRunner$$anonfun$newTask$1$$anon$4.execute(SbtRunner.scala:40)
[error]   sbt.ForkMain$Run$2.call(ForkMain.java:294)
[error]   sbt.ForkMain$Run$2.call(ForkMain.java:284)
[error]   java.util.concurrent.FutureTask.run(FutureTask.java:262)
[error]   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
[error]   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
[error]   java.lang.Thread.run(Thread.java:745)
[error] 
[error]   CAUSED BY java.lang.AbstractMethodError: model.PlanPersistenceSpec.org$specs2$matcher$FuturezBaseMatchers$_setter_$org$specs2$matcher$FuturezBaseMatchers$$DefaultDaemonThreadFactory_$eq(Ljava/util/concurrent/ThreadFactory;)V
[error]   org.specs2.matcher.FuturezBaseMatchers$class.$init$(FuturezMatchers.scala:68)
[error]   org.specs2.mutable.Specification.<init>(Specification.scala:15)
[error]   model.PlanPersistenceSpec.<init>(PlanPersistenceSpec.scala:14)
[error]   sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[error]   sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
[error]   sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[error]   java.lang.reflect.Constructor.newInstance(Constructor.java:526)
[error]   org.specs2.reflect.Classes$$anonfun$org$specs2$reflect$Classes$$createInstanceForConstructor$1$$anonfun$apply$11.apply(Classes.scala:105)
[error]   org.specs2.control.ActionT$$anonfun$safe$1.apply(ActionT.scala:89)
[error]   org.specs2.control.ActionT$$anonfun$reader$1$$anonfun$apply$6.apply(ActionT.scala:80)
[error]   org.specs2.control.Status$.safe(Status.scala:100)
[error]   org.specs2.control.StatusT$$anonfun$safe$1.apply(StatusT.scala:62)
[error]   org.specs2.control.StatusT$$anonfun$safe$1.apply(StatusT.scala:62)
[error]   scalaz.syntax.ToApplicativeOps$$anon$1.self$lzycompute(ApplicativeSyntax.scala:30)
[error]   scalaz.syntax.ToApplicativeOps$$anon$1.self(ApplicativeSyntax.scala:30)
[error]   scalaz.syntax.ToApplicativeOps$ApplicativeIdV$$anonfun$point$1.apply(ApplicativeSyntax.scala:34)
[error]   scalaz.WriterTApplicative$$anonfun$point$1.apply(WriterT.scala:266)
[error]   scalaz.WriterTApplicative$$anonfun$point$1.apply(WriterT.scala:266)
[error]   scalaz.effect.IO$$anonfun$apply$19$$anonfun$apply$20.apply(IO.scala:137)
[error]   scalaz.effect.IO$$anonfun$apply$19$$anonfun$apply$20.apply(IO.scala:137)
[error]   scalaz.FreeFunctions$$anonfun$return_$1.apply(Free.scala:288)
[error]   scalaz.FreeFunctions$$anonfun$return_$1.apply(Free.scala:288)
[error]   scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:33)
[error]   scalaz.Free$$anonfun$run$1.apply(Free.scala:174)
[error]   scalaz.Free$$anonfun$run$1.apply(Free.scala:174)
[error]   scalaz.Free.go2$1(Free.scala:117)
[error]   scalaz.Free.go(Free.scala:120)
[error]   scalaz.Free.run(Free.scala:174)
[error]   scalaz.effect.IO$class.unsafePerformIO(IO.scala:23)
[error]   scalaz.effect.IOFunctions$$anon$5.unsafePerformIO(IO.scala:221)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1$$anonfun$3.apply(Classes.scala:37)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1$$anonfun$3.apply(Classes.scala:36)
[error]   scala.collection.immutable.List.map(List.scala:273)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1.apply(Classes.scala:36)
[error]   org.specs2.reflect.Classes$$anonfun$createInstance$1$$anonfun$apply$1.apply(Classes.scala:29)
[error]   scala.Function1$$anonfun$andThen$1.apply(Function1.scala:55)
[error]   org.specs2.control.Status$class.fold(Status.scala:30)
[error]   org.specs2.control.Ok.fold(Status.scala:95)
[error]   org.specs2.control.Status$class.flatMap(Status.scala:48)
[error]   org.specs2.control.Ok.flatMap(Status.scala:95)
[error]   org.specs2.control.Status$class.map(Status.scala:45)
[error]   org.specs2.control.Ok.map(Status.scala:95)
[error]   org.specs2.control.StatusT$$anonfun$map$1.apply(StatusT.scala:16)
[error]   org.specs2.control.StatusT$$anonfun$map$1.apply(StatusT.scala:16)
[error]   scalaz.WriterT$$anonfun$map$1.apply(WriterT.scala:48)
[error]   scalaz.WriterT$$anonfun$map$1.apply(WriterT.scala:48)
[error]   scalaz.effect.IO$$anonfun$map$1$$anonfun$apply$8.apply(IO.scala:57)
[error]   scalaz.effect.IO$$anonfun$map$1$$anonfun$apply$8.apply(IO.scala:56)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:64)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:64)
[error]   scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$2.apply(Free.scala:72)
[error]   scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$2.apply(Free.scala:72)
[error]   scalaz.Free.resume(Free.scala:80)
[error]   scalaz.Free.go2$1(Free.scala:116)
[error]   scalaz.Free.go(Free.scala:120)
[error]   scalaz.Free.run(Free.scala:174)
[error]   scalaz.effect.IO$class.unsafePerformIO(IO.scala:23)
[error]   scalaz.effect.IOFunctions$$anon$5.unsafePerformIO(IO.scala:221)
[error]   org.specs2.runner.SbtRunner$$anonfun$newTask$1$$anon$4.execute(SbtRunner.scala:40)
[error]   sbt.ForkMain$Run$2.call(ForkMain.java:294)
[error]   sbt.ForkMain$Run$2.call(ForkMain.java:284)
[error]   java.util.concurrent.FutureTask.run(FutureTask.java:262)
[error]   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
[error]   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
[error]   java.lang.Thread.run(Thread.java:745)
[error] 
[error] 
[error] This looks like a specs2 exception...
[error] Please report it with the preceding stacktrace at http://github.com/etorreborre/specs2/issues
[error]  
[error] Error: Total 1, Failed 0, Errors 1, Passed 0
[error] Error during tests:
[error] 	model.PlanPersistenceSpec
[error] (test:testOnly) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 5 s, completed 13-Apr-2015 7:12:35 AM
```

As a side not, the also no longer runs in Eclipse when I run it as a JUnit test. It used to run, not sure if this is related.
```
java.lang.ClassNotFoundException: model.PlanPersistenceSpec$
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at org.specs2.reflect.Classes$$anonfun$loadClassEither$1.apply(Classes.scala:134)
	at org.specs2.reflect.Classes$$anonfun$loadClassEither$1.apply(Classes.scala:134)
	at org.specs2.control.ActionT$$anonfun$safe$1.apply(ActionT.scala:89)
	at org.specs2.control.ActionT$$anonfun$reader$1$$anonfun$apply$6.apply(ActionT.scala:80)
	at org.specs2.control.Status$.safe(Status.scala:100)
	at org.specs2.control.StatusT$$anonfun$safe$1.apply(StatusT.scala:62)
	at org.specs2.control.StatusT$$anonfun$safe$1.apply(StatusT.scala:62)
	at scalaz.syntax.ToApplicativeOps$$anon$1.self$lzycompute(ApplicativeSyntax.scala:30)
	at scalaz.syntax.ToApplicativeOps$$anon$1.self(ApplicativeSyntax.scala:30)
	at scalaz.syntax.ToApplicativeOps$ApplicativeIdV$$anonfun$point$1.apply(ApplicativeSyntax.scala:34)
	at scalaz.WriterTApplicative$$anonfun$point$1.apply(WriterT.scala:266)
	at scalaz.WriterTApplicative$$anonfun$point$1.apply(WriterT.scala:266)
	at scalaz.effect.IO$$anonfun$apply$19$$anonfun$apply$20.apply(IO.scala:137)
	at scalaz.effect.IO$$anonfun$apply$19$$anonfun$apply$20.apply(IO.scala:137)
	at scalaz.FreeFunctions$$anonfun$return_$1.apply(Free.scala:288)
	at scalaz.FreeFunctions$$anonfun$return_$1.apply(Free.scala:288)
	at scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:33)
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:174)
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:174)
	at scalaz.Free.go2$1(Free.scala:117)
	at scalaz.Free.go(Free.scala:120)
	at scalaz.Free.run(Free.scala:174)
	at scalaz.effect.IO$class.unsafePerformIO(IO.scala:23)
	at scalaz.effect.IOFunctions$$anon$5.unsafePerformIO(IO.scala:221)
	at org.specs2.runner.JUnitRunner.specification$lzycompute(JUnitRunner.scala:19)
	at org.specs2.runner.JUnitRunner.specification(JUnitRunner.scala:18)
	at org.specs2.runner.JUnitRunner.specStructure$lzycompute(JUnitRunner.scala:29)
	at org.specs2.runner.JUnitRunner.specStructure(JUnitRunner.scala:29)
	at org.specs2.runner.JUnitRunner.getDescription$lzycompute(JUnitRunner.scala:35)
	at org.specs2.runner.JUnitRunner.getDescription(JUnitRunner.scala:35)
	at org.junit.runner.Runner.testCount(Runner.java:41)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestClassReference.countTestCases(JUnit4TestClassReference.java:30)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.countTests(RemoteTestRunner.java:479)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:447)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

**Comment by etorreborre** [comment]
Can you please run a "dependency-tree" command with the sbt dependency-graph plugin to see which library are being used? Also what is the result of `show externalDependencyClasspath`? The `DefaultDaemonThreadFactory` that you see comes from `scalaz-7.1.1` and I don't see how it could have  snuck in. Maybe can you also make sure that everything is cleaned before you rebuild?

**Comment by etorreborre** [comment]
I think this is my fault. I'll try to publish a working 3.5 version ASAP.

**Comment by andyczerwonka** [comment]
Ok, I will run the deps graph as you suggest first thing tomorrow if tht helps.

**Comment by etorreborre** [comment]
Can you please try this version `3.4-scalaz-7.0.6-20150414054958-dce2bc3` and tell me if that works better for you?

**Comment by andyczerwonka** [comment]
Looks like the change you made in `3.4-scalaz-7.0.6-20150414054958-dce2bc3` works.

**Comment by etorreborre** [comment]
Yes this looks like another issue here (pretty puzzling too). I would like to investigate it but I would need access to your project (is it on Github?).

Otherwise I suggest to remove most of the moving parts here and write
```scala
@RunWith(classOf[JUnitRunner])
class MySpec extends mutable.Specification with BeforeAfterEach {
  def before = println("before")
  def after = println("after")
  "test1" >> { println("test1"); ok } 
  "test2" >> { println("test2"); ok } 
  "test3" >> { println("test3"); ok } 
  "test4" >> { println("test4"); ok } 
}
```
and see what gets printed on screen when you run in either JUnit or sbt.

**Comment by etorreborre** [comment]
BTW, it's great to see a screencast of your working environment!

**Comment by andyczerwonka** [comment]
This particular example is in a private repo, but I have a similar public repo that I'll try and reproduce the issue. Give me a couple of hours and I'll get it done.

**Comment by andyczerwonka** [comment]
Ok, so the result of that spec in sbt is as follows:
```
[freeze] $ testOnly model.SimpleSpec
[info] Compiling 1 Scala source to /home/aczerwon/work/freeze/target/scala-2.11/test-classes...
before
test1
after
before
test2
after
before
test3
after
[info] SimpleSpec
[info] 
[info] + test1
[info] + test2
[info] + test3
before
test4
after
[info] + test4
[info] 
[info] Total for specification SimpleSpec
[info] Finished in 13 ms
[info] 4 examples, 0 failure, 0 error
[info] 
[info] Passed: Total 4, Failed 0, Errors 0, Passed 4
[success] Total time: 10 s, completed 14-Apr-2015 9:41:59 PM
```
All the tests pass in Eclipse, but clearly we have a problem.
```
before
before
before
before
test4
test3
test2
test1
after
after
after
after
```

**Comment by andyczerwonka** [comment]
I created [a branch](https://github.com/andyczerwonka/tyrion/tree/spec2_issue_359) that shows the problem. If you run [SimpleSpec](https://github.com/andyczerwonka/tyrion/blob/spec2_issue_359/test/models/SimpleSpec.scala) in sbt and Eclipse (latest release version) you should see the same results.

**Comment by etorreborre** [comment]
Thanks for this. I need to install Eclipse on my machine to test your branch. I can reproduce your results in IntelliJ where the sequencing is not the same in sbt and with a JUnitRunner (but I don't see any missing tests like in your screencast, can you reproduce that on a simple spec?).

On the other hand a different trace for concurrent execution is not necessarily a problem. I need to investigate what happens when we have tests with different execution times and check that the JUnit execution doesn't force a sequential execution (in reverse order it would seem).

**Comment by etorreborre** [comment]
No, if I change the time taken by each example I get a different ordering in both sbt and JUnit.

**Comment by andyczerwonka** [comment]
Hmnnnn... I wasn't able to reproduce the missing tests with the SimpleSpec, but I'm sure I'll be able to get an example. I'll mess around in the branch and hopefully get you a working example. Signing out for tonight... bedtime! :)

**Comment by etorreborre** [comment]
Have a good night, see you tomorrow :-)

**Comment by andyczerwonka** [comment]
Ok, some interesting observations. The test that I wrote where only part of the tests complete looks like:
``` scala
@RunWith(classOf[JUnitRunner])
class PlanPersistenceSpec extends Specification with BeforeAfterEach {
  def before = db("plans").remove(MongoDBObject.empty)
  def after = db("plans").remove(MongoDBObject.empty)

  val payload = Json.parse("""{"name":"Fiver","age":4,"role":null}""")
  val now = Calendar.getInstance().getTime()
  val userId = new ObjectId("551d5cf3985f20e469a1402a")

  "The Plan DAO" should {

    "create a plan" in {
      val plan = Plan.create("Some name", userId, payload).get
      plan.id must not be null
      plan.name mustEqual "Some name"
      plan.payload mustEqual payload
    }

    "find a specific plan" in {
      val plan = create("Some name", userId, payload).get
      Plan.find(plan.id) must beSome(plan)
    }

    "not finding a specific plan" in {
      Plan.find(new ObjectId()) must beNone
    }

    "find all plans" in {
      Plan.create("A", userId, payload)
      Plan.create("B", userId, payload)
      Plan.findAllByUser(userId) must have size (2)
    }

    "update a plan" in {
      val plan = Plan.create("A", userId, payload).get
      Plan.find(plan.id) must beSome(plan)
      val updated = plan.copy(name = "B")
      val result = Plan.update(updated)
      result must beSuccessfulTry.withValue(updated)
      Plan.find(plan.id) must beSome(updated)
    }

    "delete a plan" in {
      val id = Plan.create("A", userId, payload).get.id
      Plan.findAllByUser(userId) must have size (1)
      Plan.delete(id) must beSuccessfulTry
      Plan.findAllByUser(userId) must be empty
    }

    "delete a non existant plan does nothing" in {
      val json = Plan.create("A", userId, payload)
      Plan.delete(new ObjectId()) must beSuccessfulTry
      Plan.findAllByUser(userId) must have size (1)
    }

    "create a plan with empty playload" in {
      val plan = Plan.create("Some name", userId, Json.parse("""{}""")).get
      plan.id must not be null
      plan.name mustEqual "Some name"
      Json.stringify(plan.payload) mustEqual """{}"""
    }

  }

}
```
The test runs in SBT just fine, but when I run it in JUnit with Eclipse I get 3, maybe 4 of the test executed and they fail. I think the test is inherently wrong in that each test tries to clean the database, but because the tests are running concurrently, who knows what's in the collection. To test that theory, I added `sequential` to this test and it ran correctly in JUnit. So the question is, why do the tests always success in sbt? There is clearly different behaviour in the runners.

You can reproduce this behaviour by running [this test](https://github.com/andyczerwonka/tyrion/blob/spec2_issue_359/test/models/UserPersistenceSpec.scala). Run the test in sbt and JUnit and you will see it pass. Remove the `sequential` and it will still run correctly in sbt but will fail in JUnit.

**Note:** The test depends on MongoDB, but if you install vanilla 2.8 with all the defaults it will run. No configuration necessary.

**Comment by etorreborre** [comment]
I want to make sure I understand your context. You are running a specification against a MongoDB database that is shared across examples. Before / After each example you drop the database. I actually think that if the specification is executed concurrently you *should* see failures because some examples will try to execute when others will have already written some data in the db (and you'll get failures about duplicate keys or collections sizes).

In that case the question is more: why does this not happen with the sbt run? It seems that the default concurrent execution model is broken with specs2 when using scalaz 7.0.6 on your project (I can't reproduce the issue in specs2 itself when using scalaz 7.0.6 for now).

Anyway this also means that you *have* to run your specification sequentially or you have to design your examples so that they don't interact with each other when using the same database.

**Comment by andyczerwonka** [comment]
That's correct. I now feel like I should be seeing errors, and I do under JUnit. sbt doesn't seem to run concurrently.

**Comment by etorreborre** [comment]
That's still a mystery to me. I tried various combinations of specs2 3.4, 2.4.17 with scalaz 7.0.6 or scalaz 7.1.1 and I get the same issue. Maybe the play plugin is doing something weird here. I suggest that we close this issue and you reopen one on Play's site.

**Comment by andyczerwonka** [comment]
When you say 'same issue', are you saying it works correctly for you? Consistently across sbt and JUnit?

**Comment by etorreborre** [comment]
"same issue" means that the spec is executed sequentially with sbt on your project with whatever version of specs2 I use (and it works fine in a non-Play project).

**Comment by andyczerwonka** [comment]
wow, that's strange. Sequential in Play's activator and concurrent in native sbt. Bizarre.

**Comment by etorreborre** [comment]
I created playframework/playframework#4253 to track this issue with the Play's team.

**Comment by andyczerwonka** [comment]
Thanks Eric. I take it the fix you made will make the 3.5 release?

**Comment by etorreborre** [comment]
Yes definitely.
